README
#
# (C) Copyright 2000 - 2005
# Wolfgang Denk, DENX Software Engineering, wd@denx.de.
#
# See file CREDITS for list of people who contributed to this
# project.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of
# the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA 02111-1307 USA
#

Summary:
========

This directory contains the source code for U-Boot, a boot loader for
Embedded boards based on PowerPC, ARM, MIPS and several other
processors, which can be installed in a boot ROM and used to
initialize and test the hardware or to download and run application
code.

The development of U-Boot is closely related to Linux: some parts of
the source code originate in the Linux source tree, we have some
header files in common, and special provision has been made to
support booting of Linux images.

Some attention has been paid to make this software easily
configurable and extendable. For instance, all monitor commands are
implemented with the same call interface, so that it's very easy to
add new commands. Also, instead of permanently adding rarely used
code (for instance hardware test utilities) to the monitor, you can
load and run it dynamically.


Status:
=======

In general, all boards for which a configuration option exists in the
Makefile have been tested to some extent and can be considered
"working". In fact, many of them are used in production systems.

In case of problems see the CHANGELOG and CREDITS files to find out
who contributed the specific port.


Where to get help:
==================

In case you have questions about, problems with or contributions for
U-Boot you should send a message to the U-Boot mailing list at
<u-boot-users@lists.sourceforge.net>. There is also an archive of
previous traffic on the mailing list - please search the archive
before asking FAQ's. Please see
http://lists.sourceforge.net/lists/listinfo/u-boot-users/


Where we come from:
===================

- start from 8xxrom sources
- create PPCBoot project (http://sourceforge.net/projects/ppcboot)
- clean up code
- make it easier to add custom boards
- make it possible to add other [PowerPC] CPUs
- extend functions, especially:
  * Provide extended interface to Linux boot loader
  * S-Record download
  * network boot
  * PCMCIA / CompactFLash / ATA disk / SCSI ... boot
- create ARMBoot project (http://sourceforge.net/projects/armboot)
- add other CPU families (starting with ARM)
- create U-Boot project (http://sourceforge.net/projects/u-boot)


Names and Spelling:
===================

The "official" name of this project is "Das U-Boot". The spelling
"U-Boot" shall be used in all written text (documentation, comments
in source files etc.). Example:

	This is the README file for the U-Boot project.

File names etc. shall be based on the string "u-boot". Examples:

	include/asm-ppc/u-boot.h

	#include <asm/u-boot.h>

Variable names, preprocessor constants etc. shall be either based on
the string "u_boot" or on "U_BOOT". Example:

	U_BOOT_VERSION		u_boot_logo
	IH_OS_U_BOOT		u_boot_hush_start


Versioning:
===========

U-Boot uses a 3 level version number containing a version, a
sub-version, and a patchlevel: "U-Boot-2.34.5" means version "2",
sub-version "34", and patchlevel "4".

The patchlevel is used to indicate certain stages of development
between released versions, i. e. officially released versions of
U-Boot will always have a patchlevel of "0".


Directory Hierarchy:
====================

- board		Board dependent files
- common	Misc architecture independent functions
- cpu		CPU specific files
  - 74xx_7xx	Files specific to Freescale MPC74xx and 7xx CPUs
  - arm720t	Files specific to ARM 720 CPUs
  - arm920t	Files specific to ARM 920 CPUs
    - at91rm9200 Files specific to Atmel AT91RM9200 CPU
    - imx	Files specific to Freescale MC9328 i.MX CPUs
    - s3c24x0	Files specific to Samsung S3C24X0 CPUs
  - arm925t	Files specific to ARM 925 CPUs
  - arm926ejs	Files specific to ARM 926 CPUs
  - arm1136	Files specific to ARM 1136 CPUs
  - i386	Files specific to i386 CPUs
  - ixp		Files specific to Intel XScale IXP CPUs
  - mcf52x2	Files specific to Freescale ColdFire MCF52x2 CPUs
  - mips	Files specific to MIPS CPUs
  - mpc5xx	Files specific to Freescale MPC5xx  CPUs
  - mpc5xxx	Files specific to Freescale MPC5xxx CPUs
  - mpc8xx	Files specific to Freescale MPC8xx  CPUs
  - mpc8220	Files specific to Freescale MPC8220 CPUs
  - mpc824x	Files specific to Freescale MPC824x CPUs
  - mpc8260	Files specific to Freescale MPC8260 CPUs
  - mpc85xx	Files specific to Freescale MPC85xx CPUs
  - nios	Files specific to Altera NIOS CPUs
  - nios2	Files specific to Altera Nios-II CPUs
  - ppc4xx	Files specific to AMCC PowerPC 4xx CPUs
  - pxa		Files specific to Intel XScale PXA CPUs
  - s3c44b0	Files specific to Samsung S3C44B0 CPUs
  - sa1100	Files specific to Intel StrongARM SA1100 CPUs
- disk		Code for disk drive partition handling
- doc		Documentation (don't expect too much)
- drivers	Commonly used device drivers
- dtt		Digital Thermometer and Thermostat drivers
- examples	Example code for standalone applications, etc.
- include	Header Files
- lib_arm	Files generic to ARM	 architecture
- lib_generic	Files generic to all	 architectures
- lib_i386	Files generic to i386	 architecture
- lib_m68k	Files generic to m68k	 architecture
- lib_mips	Files generic to MIPS	 architecture
- lib_nios	Files generic to NIOS	 architecture
- lib_ppc	Files generic to PowerPC architecture
- net		Networking code
- post		Power On Self Test
- rtc		Real Time Clock drivers
- tools		Tools to build S-Record or U-Boot images, etc.

Software Configuration:
=======================

Configuration is usually done using C preprocessor defines; the
rationale behind that is to avoid dead code whenever possible.

There are two classes of configuration variables:

* Configuration _OPTIONS_:
  These are selectable by the user and have names beginning with
  "CONFIG_".

* Configuration _SETTINGS_:
  These depend on the hardware etc. and should not be meddled with if
  you don't know what you're doing; they have names beginning with
  "CFG_".

Later we will add a configuration tool - probably similar to or even
identical to what's used for the Linux kernel. Right now, we have to
do the configuration by hand, which means creating some symbolic
links and editing some configuration files. We use the TQM8xxL boards
as an example here.


Selection of Processor Architecture and Board Type:
---------------------------------------------------

For all supported boards there are ready-to-use default
configurations available; just type "make <board_name>_config".

Example: For a TQM823L module type:

	cd u-boot
	make TQM823L_config

For the Cogent platform, you need to specify the cpu type as well;
e.g. "make cogent_mpc8xx_config". And also configure the cogent
directory according to the instructions in cogent/README.


Configuration Options:
----------------------

Configuration depends on the combination of board and CPU type; all
such information is kept in a configuration file
"include/configs/<board_name>.h".

Example: For a TQM823L module, all configuration settings are in
"include/configs/TQM823L.h".


Many of the options are named exactly as the corresponding Linux
kernel configuration options. The intention is to make it easier to
build a config tool - later.


The following options need to be configured:

- CPU Type:	Define exactly one of

		PowerPC based CPUs:
		-------------------
		CONFIG_MPC823,	CONFIG_MPC850,	CONFIG_MPC855,	CONFIG_MPC860
	or	CONFIG_MPC5xx
	or	CONFIG_MPC8220
	or	CONFIG_MPC824X, CONFIG_MPC8260
	or	CONFIG_MPC85xx
	or	CONFIG_IOP480
	or	CONFIG_405GP
	or	CONFIG_405EP
	or	CONFIG_440
	or	CONFIG_MPC74xx
	or	CONFIG_750FX

		ARM based CPUs:
		---------------
		CONFIG_SA1110
		CONFIG_ARM7
		CONFIG_PXA250
		CONFIG_CPU_MONAHANS

		MicroBlaze based CPUs:
		----------------------
		CONFIG_MICROBLAZE

		Nios-2 based CPUs:
		----------------------
		CONFIG_NIOS2


- Board Type:	Define exactly one of

		PowerPC based boards:
		---------------------

		CONFIG_ADCIOP		CONFIG_FPS860L		CONFIG_OXC
		CONFIG_ADS860		CONFIG_GEN860T		CONFIG_PCI405
		CONFIG_AMX860		CONFIG_GENIETV		CONFIG_PCIPPC2
		CONFIG_AP1000		CONFIG_GTH		CONFIG_PCIPPC6
		CONFIG_AR405		CONFIG_gw8260		CONFIG_pcu_e
		CONFIG_BAB7xx		CONFIG_hermes		CONFIG_PIP405
		CONFIG_BC3450		CONFIG_hymod		CONFIG_PM826
		CONFIG_c2mon		CONFIG_IAD210		CONFIG_ppmc8260
		CONFIG_CANBT		CONFIG_ICU862		CONFIG_QS823
		CONFIG_CCM		CONFIG_IP860		CONFIG_QS850
		CONFIG_CMI		CONFIG_IPHASE4539	CONFIG_QS860T
		CONFIG_cogent_mpc8260	CONFIG_IVML24		CONFIG_RBC823
		CONFIG_cogent_mpc8xx	CONFIG_IVML24_128	CONFIG_RPXClassic
		CONFIG_CPCI405		CONFIG_IVML24_256	CONFIG_RPXlite
		CONFIG_CPCI4052		CONFIG_IVMS8		CONFIG_RPXsuper
		CONFIG_CPCIISER4	CONFIG_IVMS8_128	CONFIG_rsdproto
		CONFIG_CPU86		CONFIG_IVMS8_256	CONFIG_sacsng
		CONFIG_CRAYL1		CONFIG_JSE		CONFIG_Sandpoint8240
		CONFIG_CSB272		CONFIG_LANTEC		CONFIG_Sandpoint8245
		CONFIG_CU824		CONFIG_LITE5200B	CONFIG_sbc8260
		CONFIG_DASA_SIM		CONFIG_lwmon		CONFIG_sbc8560
		CONFIG_DB64360		CONFIG_MBX		CONFIG_SM850
		CONFIG_DB64460		CONFIG_MBX860T		CONFIG_SPD823TS
		CONFIG_DU405		CONFIG_MHPC		CONFIG_STXGP3
		CONFIG_DUET_ADS		CONFIG_MIP405		CONFIG_SXNI855T
		CONFIG_EBONY		CONFIG_MOUSSE		CONFIG_TQM823L
		CONFIG_ELPPC		CONFIG_MPC8260ADS	CONFIG_TQM8260
		CONFIG_ELPT860		CONFIG_MPC8540ADS	CONFIG_TQM850L
		CONFIG_ep8260		CONFIG_MPC8540EVAL	CONFIG_TQM855L
		CONFIG_ERIC		CONFIG_MPC8560ADS	CONFIG_TQM860L
		CONFIG_ESTEEM192E	CONFIG_MUSENKI		CONFIG_TTTech
		CONFIG_ETX094		CONFIG_MVS1		CONFIG_UTX8245
		CONFIG_EVB64260		CONFIG_NETPHONE		CONFIG_V37
		CONFIG_FADS823		CONFIG_NETTA		CONFIG_W7OLMC
		CONFIG_FADS850SAR	CONFIG_NETVIA		CONFIG_W7OLMG
		CONFIG_FADS860T		CONFIG_NX823		CONFIG_WALNUT
		CONFIG_FLAGADM		CONFIG_OCRTC		CONFIG_ZPC1900
		CONFIG_FPS850L		CONFIG_ORSG		CONFIG_ZUMA

		ARM based boards:
		-----------------

		CONFIG_ARMADILLO,	CONFIG_AT91RM9200DK,	CONFIG_CERF250,
		CONFIG_CSB637,		CONFIG_DELTA,		CONFIG_DNP1110,
		CONFIG_EP7312,		CONFIG_H2_OMAP1610,	CONFIG_HHP_CRADLE,

CHANGELOG
======================================================================
Changes since U-Boot 1.1.4:
======================================================================

* Add port of FAT filesystem driver from the Rockbox (www.rockbox.org)
  project.  Driver provides a "roughly POSIX-compatible file IO API"
  including read, write, open, close and ls.
  Patch by Keith Outwater, 04 May 2006

* Prevent USB commands from working when USB is stopped.

* Add rudimentary handling of alternate settings of USB interfaces.
  This is in order to fix issues with some USB sticks timing out
  during initialization. Some code readability improvements.

* PPC440 DDR setup: Set SDRAM0_CFG0[PMU]=0 for best performance
  AMCC suggested to set the PMU bit to 0 for best performace on
  the PPC440 DDR controller.
  Please see doc/README.440-DDR-performance for details.
  Patch by Stefan Roese, 28 Jul 2006

* AMCC bamboo (440EP) U-Boot image reduced to 384kbyte
  Please see doc/README.bamboo for details.
  Patch by Stefan Roese, 27 Jul 2006

* Fix CONFIG_CMDLINE_EDITING implementation
  Patch by Stefan Roese, 27 Jul 2006

* Fix preboot message on TQM5200 after switching to hush parser.

* MCC200: set default configuration to low_boot DDR,
  and support for configurable options high_boot and/or SDRAM.

* Add support for 256 MB SDRAM on CPU87
  Patch by Josef Wagner, 25 Nov 2005

* Add configuration for cam5200 board (based on TQM5200S).

* More code cleanup

* Disabled kvme080 board in MAKEALL because of build problems.

* Code cleanup

* Update NetStar board
  Patch by Ladislav Michl, 03 Nov 2005

* Make code better readable.
  Patch by Ladislav Michl, 14 Sep 2005

* Enable initrd ATAG for xm250 board.
  Patch by Josef Wagner, 05 Sep 2005

* Add readline cmdline-editing extension
  Patch by JinHua Luo, 01 Sep 2005

* Add support for friendly-arm SBC-2410X board
  Patch by JinHua Luo, 01 Sep 2005

* Fix multi-part image support on i386 platform.
  Patch by David Updegraff, 19 Aug 2005

* Add support for KVME080 board
  Patch by Sangmoon Kim, 18 Aug 2005

* Fix MIPS LE build problem
  Patch by Matej Kupljen, 10 Aug 2005

* Check argument count in "mii" command.
  Problem pointed out by Andrew Dyer, 13 Jun 2005

* Cleanup TQM5200 board configurations:
  - make highboot configurations use environment at high end, too,
    to avoid flash fragmentation
  - always use redundand environment
  - don't enable video code for modules without graphics controller
  - provide useful (though different) mtdparts settings
  - get rid of CONFIG_CS_AUTOCONF which was always set anyway

* Extend mkconfig tool to print more useful target name

* Add support for high-boot on TQM5200 and TQM5200S boards.
  Hint: the CPLD on the TQM5200 must be programmed with a software
  version supporting the high boot option! The new TQM5200S is
  already supporting this option. On the TQM5200 this option will be
  supported in configurations with MPC5200 rev B processors.
  To actually "high boot", set jumper X30 on the STK52xx.
  Patch by Martin Krause, 12 Jul 2006

* Add support for new TQM5200 revisions
  - Support for TQM5200S (short version without graphic controller)
  - Support for modules with 'N' type S29GL128N Spansion flashes
    (requires changes to flash layout)
  - Support for MPC5200B cpu (mostly support for second SDRAM bank)
  Patch by Martin Krause, 07 Jul 2006

* Fix support for PS/2 keyboard on TQM85xx boards
  The PS/2 keyobard driver for the TQM85xx modules only supports the
  internal DUART of the MPC85xx CPU. Since the MPC8560 doesn't
  include a DUART, the TQM8560 modules can't be used with the PS/2
  keyboard controller on the STK85xx board.
  The PS/2 keyboard driver should work with the modules TQM8540,
  TQM8541 and TQM8555, but it only has been tested on a TQM8540, yet.
  Make sure the PS/2 controller on the STK85xx is programmed. Jumper
  settings: X66 1-2, 9-10; X61 2-3
  Patch by Martin Krause, 21 Jun 2006

* Adjust RTC century handling on STK52xx board to match Linux driver.
  Patch by Martin Krause, 12 Jun 2006

* Adjust filenames for USB update images on TRAB board.
  During an automatic update via USB stick, U-Boot searches for
  images with the name "firmware.img" and "kernel.img". This names
  are now changed to "firmw_01.img" and "kernl_01.img". This is done,
  to prevent updates of new boards (with the new macronics "c" step
  flashes) with old, incompatible firmware or kernel versions.
  Patch by Martin Krause, 21 Jun 2006

* Bugfix in VFD routine on TRAB board.
  Make sure upper lext pixel can be set to blue, too
  (so far only red was possible).
  Patch by Martin Krause, 15 Feb 2006

* Enable buffered flash writes for TB5200 board.

* Fix some bugs in TRAB board flash driver.
  - increase CFG_FLASH_ERASE_TOUT from 2 to 15 seconds
  - use CFG_FLASH_WRITE_TOUT for programming instead of CFG_FLASH_ERASE_TOUT
  - remove "Unlock Bypass" mode, because macronix flashes do not support
    this mode officially
  - fix flash reset command from 0x00FF to 0x00F0. 0x00FF is only specified
    for Intel compatible flashes, not for AMD compatible.
  Patch by Martin Krause, 15 Feb 2006

* Add additional error messages to flash driver on TRAB board
  (for erase errors and timeout errors)
  Patch by Martin Krause, 14 Feb 2006

* Add support for TB5200 board

lib_arm/board.c
/*
 * (C) Copyright 2002-2006
 * Wolfgang Denk, DENX Software Engineering, wd@denx.de.
 *
 * (C) Copyright 2002
 * Sysgo Real-Time Solutions, GmbH <www.elinos.com>
 * Marius Groeger <mgroeger@sysgo.de>
 *
 * See file CREDITS for list of people who contributed to this
 * project.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of
 * the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston,
 * MA 02111-1307 USA
 */

/*
 * To match the U-Boot user interface on ARM platforms to the U-Boot
 * standard (as on PPC platforms), some messages with debug character
 * are removed from the default U-Boot build.
 *
 * Define DEBUG here if you want additional info as shown below
 * printed upon startup:
 *
 * U-Boot code: 00F00000 -> 00F3C774  BSS: -> 00FC3274
 * IRQ Stack: 00ebff7c
 * FIQ Stack: 00ebef7c
 */

#include <common.h>
#include <command.h>
#include <malloc.h>
#include <devices.h>
#include <version.h>
#include <net.h>

#if defined(CONFIG_LCD_ILI9327) || defined(BOARD_WITH_LCD_SHOW)
#include <asm/arch/dss.h>
#endif

#ifdef CONFIG_DRIVER_SMC91111
#include "../drivers/smc91111.h"
#endif
#ifdef CONFIG_DRIVER_LAN91C96
#include "../drivers/lan91c96.h"
#endif

#if defined(CONFIG_BOOTCASE_CHECK)
#include <twl4030.h>
#endif

DECLARE_GLOBAL_DATA_PTR;

#if (CONFIG_COMMANDS & CFG_CMD_NAND)
#ifdef ENV_IS_VARIABLE
extern u8 is_nand;
#endif
void nand_init (void);
#endif

#if (CONFIG_COMMANDS & CFG_CMD_FLASH)
#ifdef ENV_IS_VARIABLE
extern u8 is_flash;
#endif
#endif

#if (CONFIG_COMMANDS & CFG_CMD_ONENAND)
#ifdef ENV_IS_VARIABLE
extern u8 is_onenand;
#endif
void onenand_init(void);
#endif

ulong monitor_flash_len;

#ifdef CONFIG_HAS_DATAFLASH
extern int  AT91F_DataflashInit(void);
extern void dataflash_print_info(void);
#endif

#if (CONFIG_STORAGE_EMMC)
extern void board_mmc_init(void);
#endif

#ifndef CONFIG_IDENT_STRING
#define CONFIG_IDENT_STRING ""
#endif

const char version_string[] =
	U_BOOT_VERSION" (" __DATE__ " - " __TIME__ ")"CONFIG_IDENT_STRING;

#ifdef CONFIG_DRIVER_CS8900
extern void cs8900_get_enetaddr (uchar * addr);
#endif

#ifdef CONFIG_DRIVER_RTL8019
extern void rtl8019_get_enetaddr (uchar * addr);
#endif

/*
 * Begin and End of memory area for malloc(), and current "brk"
 */
static ulong mem_malloc_start = 0;
static ulong mem_malloc_end = 0;
static ulong mem_malloc_brk = 0;

static
void mem_malloc_init (ulong dest_addr)
{
	mem_malloc_start = dest_addr;
	mem_malloc_end = dest_addr + CFG_MALLOC_LEN;
	mem_malloc_brk = mem_malloc_start;

	memset ((void *) mem_malloc_start, 0,
			mem_malloc_end - mem_malloc_start);
}

void *sbrk (ptrdiff_t increment)
{
	ulong old = mem_malloc_brk;
	ulong new = old + increment;

	if ((new < mem_malloc_start) || (new > mem_malloc_end)) {
		return (NULL);
	}
	mem_malloc_brk = new;

	return ((void *) old);
}

/************************************************************************
 * Init Utilities							*
 ************************************************************************
 * Some of this code should be moved into the core functions,
 * or dropped completely,
 * but let's get it working (again) first...
 */

static int init_baudrate (void)
{
	char tmp[64];	/* long enough for environment variables */
	int i = getenv_r ("baudrate", tmp, sizeof (tmp));
	gd->bd->bi_baudrate = gd->baudrate = (i > 0)
			? (int) simple_strtoul (tmp, NULL, 10)
			: CONFIG_BAUDRATE;

	return (0);
}

static int display_banner (void)
{
	printf ("\n\n%s\n\n", version_string);
	debug ("U-Boot code: %08lX -> %08lX  BSS: -> %08lX\n",
	       _armboot_start, _bss_start, _bss_end);
#ifdef CONFIG_MODEM_SUPPORT
	debug ("Modem Support enabled\n");
#endif
#ifdef CONFIG_USE_IRQ
	debug ("IRQ Stack: %08lx\n", IRQ_STACK_START);
	debug ("FIQ Stack: %08lx\n", FIQ_STACK_START);
#endif

	return (0);
}

/*
 * WARNING: this code looks "cleaner" than the PowerPC version, but
 * has the disadvantage that you either get nothing, or everything.
 * On PowerPC, you might see "DRAM: " before the system hangs - which
 * gives a simple yet clear indication which part of the
 * initialization if failing.
 */
static int display_dram_config (void)
{
	int i;

#ifdef DEBUG
	puts ("RAM Configuration:\n");

	for(i=0; i<CONFIG_NR_DRAM_BANKS; i++) {
		printf ("Bank #%d: %08lx ", i, gd->bd->bi_dram[i].start);
		print_size (gd->bd->bi_dram[i].size, "\n");
	}
#else
	ulong size = 0;

	for (i=0; i<CONFIG_NR_DRAM_BANKS; i++) {
		size += gd->bd->bi_dram[i].size;
	}
	puts("DRAM:  ");
	print_size(size, "\n");
#endif

	return (0);
}

#ifndef CFG_NO_FLASH
static void display_flash_config (ulong size)
{
	puts ("Flash: ");
	print_size (size, "\n");
}
#endif /* CFG_NO_FLASH */


/*
 * Breathe some life into the board...
 *
 * Initialize a serial port as console, and carry out some hardware
 * tests.
 *
 * The first part of initialization is running from Flash memory;
 * its main purpose is to initialize the RAM so that we
 * can relocate the monitor code to RAM.
 */

/*
 * All attempts to come up with a "common" initialization sequence
 * that works for all boards and architectures failed: some of the
 * requirements are just _too_ different. To get rid of the resulting
 * mess of board dependent #ifdef'ed code we now make the whole
 * initialization sequence configurable to the user.
 *
 * The requirements for any new initalization function is simple: it
 * receives a pointer to the "global data" structure as it's only
 * argument, and returns an integer return code, where 0 means
 * "continue" and != 0 means "fatal error, hang the system".
 */
typedef int (init_fnc_t) (void);

int print_cpuinfo (void); /* test-only */

init_fnc_t *init_sequence[] = {
	cpu_init,		/* basic cpu dependent setup */
	board_init,		/* basic board dependent setup */
	interrupt_init,		/* set up exceptions */
	env_init,		/* initialize environment */
	init_baudrate,		/* initialze baudrate settings */
	serial_init,		/* serial communications setup */
	console_init_f,		/* stage 1 init of console */
	display_banner,		/* say that we are here */
#if defined(CONFIG_DISPLAY_CPUINFO)
	print_cpuinfo,		/* display cpu info (and speed) */
#endif
#if defined(CONFIG_DISPLAY_BOARDINFO)
	checkboard,		/* display board info */
#endif
	dram_init,		/* configure available RAM banks */
	display_dram_config,
	NULL,
};

void start_armboot (void)
{
	init_fnc_t **init_fnc_ptr;
	char *s;
#ifndef CFG_NO_FLASH
	ulong size = 0;
#endif
#if defined(CONFIG_VFD) || defined(CONFIG_LCD)
	unsigned long addr;
#endif

	/* Pointer is writable since we allocated a register for it */
	gd = (gd_t*)(_armboot_start - CFG_MALLOC_LEN - sizeof(gd_t));
	/* compiler optimization barrier needed for GCC >= 3.4 */
	__asm__ __volatile__("": : :"memory");

	memset ((void*)gd, 0, sizeof (gd_t));
	gd->bd = (bd_t*)((char*)gd - sizeof(bd_t));
	memset (gd->bd, 0, sizeof (bd_t));

	monitor_flash_len = _bss_start - _armboot_start;

	for (init_fnc_ptr = init_sequence; *init_fnc_ptr; ++init_fnc_ptr) {
		if ((*init_fnc_ptr)() != 0) {
			hang ();
		}
	}

#if defined(CONFIG_BOOTCASE_CHECK)
	/*check the boot case, add "bootcase= " in commnand line */
	bootcase_check();
#endif
        vib_ctrl(1);          //jhy add for vib ctrl  0915
        udelay(100000);
        vib_ctrl(0);

#if defined(CONFIG_LCD_ILI9327) || defined(BOARD_WITH_LCD_SHOW)
    /*initialize LCD*/
    kunlun_lcd_init();
    kunlun_lcd_display_logo();
#endif /* CONFIG_LCD_ILI9327 */

#ifndef CFG_NO_FLASH
	/* configure available FLASH banks */
#ifdef ENV_IS_VARIABLE
	if (is_flash) 
#endif
	size = flash_init ();
	display_flash_config (size);
#endif /* CFG_NO_FLASH */

#ifdef CONFIG_VFD
#	ifndef PAGE_SIZE
#	  define PAGE_SIZE 4096
#	endif
	/*
	 * reserve memory for VFD display (always full pages)
	 */
	/* bss_end is defined in the board-specific linker script */
	addr = (_bss_end + (PAGE_SIZE - 1)) & ~(PAGE_SIZE - 1);
	size = vfd_setmem (addr);
	gd->fb_base = addr;
#endif /* CONFIG_VFD */

#ifdef CONFIG_LCD
#	ifndef PAGE_SIZE
#	  define PAGE_SIZE 4096
#	endif
	/*
	 * reserve memory for LCD display (always full pages)
	 */
	/* bss_end is defined in the board-specific linker script */
	addr = (_bss_end + (PAGE_SIZE - 1)) & ~(PAGE_SIZE - 1);
	size = lcd_setmem (addr);
	gd->fb_base = addr;
#endif /* CONFIG_LCD */

	/* armboot_start is defined in the board-specific linker script */
	mem_malloc_init (_armboot_start - CFG_MALLOC_LEN);

#if (CONFIG_COMMANDS & CFG_CMD_NAND)
#ifdef ENV_IS_VARIABLE
	if (is_nand) 
#endif
	{
		puts ("NAND:");
		nand_init();		/* go init the NAND */
	}
#endif

#if (CONFIG_COMMANDS & CFG_CMD_ONENAND)
#ifdef ENV_IS_VARIABLE
	if (is_onenand)
#endif
	onenand_init();
#endif

#ifdef CONFIG_HAS_DATAFLASH
	AT91F_DataflashInit();
	dataflash_print_info();
#endif

#if (CONFIG_STORAGE_EMMC)
	/* puts("EMMC partitons init:"); */
	board_mmc_init();
#endif

	/* initialize environment */
	env_relocate ();
/*check bootcase name if download it will do fastboot*/
extern char *bootcase_name;
	if(!strcmp(bootcase_name, "download")){
		printf("strcmp(bootcase_name, download\n");
		do_fastboot(NULL, 1, 1, NULL);
	}

#ifdef CONFIG_VFD
	/* must do this after the framebuffer is allocated */
	drv_vfd_init();
#endif /* CONFIG_VFD */

	/* IP Address */
	gd->bd->bi_ip_addr = getenv_IPaddr ("ipaddr");

	/* MAC Address */
	{
		int i;
		ulong reg;
		char *s, *e;
		char tmp[64];

		i = getenv_r ("ethaddr", tmp, sizeof (tmp));
		s = (i > 0) ? tmp : NULL;

		for (reg = 0; reg < 6; ++reg) {
			gd->bd->bi_enetaddr[reg] = s ? simple_strtoul (s, &e, 16) : 0;
			if (s)
				s = (*e) ? e + 1 : e;
		}

#ifdef CONFIG_HAS_ETH1
		i = getenv_r ("eth1addr", tmp, sizeof (tmp));
		s = (i > 0) ? tmp : NULL;

		for (reg = 0; reg < 6; ++reg) {
			gd->bd->bi_enet1addr[reg] = s ? simple_strtoul (s, &e, 16) : 0;
			if (s)
				s = (*e) ? e + 1 : e;
		}
#endif
	}

	devices_init ();	/* get the devices list going. */

#ifdef CONFIG_CMC_PU2
	load_sernum_ethaddr ();
#endif /* CONFIG_CMC_PU2 */

	jumptable_init ();

	console_init_r ();	/* fully init console as a device */

#if defined(CONFIG_MISC_INIT_R)
	/* miscellaneous platform dependent initialisations */
	misc_init_r ();
#endif

	/* enable exceptions */
	enable_interrupts ();

	/* Perform network card initialisation if necessary */
#ifdef CONFIG_DRIVER_CS8900
	cs8900_get_enetaddr (gd->bd->bi_enetaddr);
#endif

#if defined(CONFIG_DRIVER_SMC91111) || defined (CONFIG_DRIVER_LAN91C96)
	if (getenv ("ethaddr")) {
		smc_set_mac_addr(gd->bd->bi_enetaddr);
	}
#endif /* CONFIG_DRIVER_SMC91111 || CONFIG_DRIVER_LAN91C96 */

	/* Initialize from environment */
	if ((s = getenv ("loadaddr")) != NULL) {
		load_addr = simple_strtoul (s, NULL, 16);
	}
#if (CONFIG_COMMANDS & CFG_CMD_NET)
	if ((s = getenv ("bootfile")) != NULL) {
		copy_filename (BootFile, s, sizeof (BootFile));
	}
#endif	/* CFG_CMD_NET */

#ifdef BOARD_LATE_INIT
	board_late_init ();
#endif
#if (CONFIG_COMMANDS & CFG_CMD_NET)
#if defined(CONFIG_NET_MULTI)
	puts ("Net:   ");
#endif
	eth_initialize(gd->bd);
#endif
	/* main_loop() can return to retry autoboot, if so just run it again. */
	for (;;) {
		main_loop ();
	}

	/* NOTREACHED - no way out of command loop except booting */
}

void hang (void)
{
	puts ("### ERROR ### Please RESET the board ###\n");
	for (;;);
}

#ifdef CONFIG_MODEM_SUPPORT
static inline void mdm_readline(char *buf, int bufsiz);

/* called from main loop (common/main.c) */
extern void  dbg(const char *fmt, ...);
int mdm_init (void)
{
	char env_str[16];
	char *init_str;
	int i;
	extern char console_buffer[];
	extern void enable_putc(void);
	extern int hwflow_onoff(int);

	enable_putc(); /* enable serial_putc() */

#ifdef CONFIG_HWFLOW
	init_str = getenv("mdm_flow_control");
	if (init_str && (strcmp(init_str, "rts/cts") == 0))
		hwflow_onoff (1);
	else
		hwflow_onoff(-1);
#endif

	for (i = 1;;i++) {
		sprintf(env_str, "mdm_init%d", i);
		if ((init_str = getenv(env_str)) != NULL) {
			serial_puts(init_str);
			serial_puts("\n");
			for(;;) {
				mdm_readline(console_buffer, CFG_CBSIZE);
				dbg("ini%d: [%s]", i, console_buffer);

				if ((strcmp(console_buffer, "OK") == 0) ||
					(strcmp(console_buffer, "ERROR") == 0)) {
					dbg("ini%d: cmd done", i);
					break;
				} else /* in case we are originating call ... */
					if (strncmp(console_buffer, "CONNECT", 7) == 0) {
						dbg("ini%d: connect", i);
						return 0;
					}
			}
		} else
			break; /* no init string - stop modem init */

		udelay(100000);
	}

	udelay(100000);

	/* final stage - wait for connect */
	for(;i > 1;) { /* if 'i' > 1 - wait for connection
				  message from modem */
		mdm_readline(console_buffer, CFG_CBSIZE);
		dbg("ini_f: [%s]", console_buffer);
		if (strncmp(console_buffer, "CONNECT", 7) == 0) {
			dbg("ini_f: connected");
			return 0;
		}
	}

	return 0;
}

/* 'inline' - We have to do it fast */
static inline void mdm_readline(char *buf, int bufsiz)
{
	char c;
	char *p;
	int n;

	n = 0;
	p = buf;
	for(;;) {
		c = serial_getc();

		/*		dbg("(%c)", c); */

		switch(c) {
		case '\r':
			break;
		case '\n':
			*p = '\0';
			return;

		default:
			if(n++ > bufsiz) {
				*p = '\0';
				return; /* sanity check */
			}
			*p = c;
			p++;
			break;
		}
	}
}
#endif	/* CONFIG_MODEM_SUPPORT */
